File: src/uk/yermak/audiobookconverter/FFMpegConcatenator.java
package uk.yermak.audiobookconverter;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.output.NullOutputStream;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * Created by Yermak on 29-Dec-17.
 */
public class FFMpegConcatenator implements Concatenator {

    private final List<Future<ConverterOutput>> futures;
    private final String outputFileName;

    public FFMpegConcatenator(List<Future<ConverterOutput>> futures, String outputFileName) {
        this.futures = futures;
        this.outputFileName = outputFileName;
    }

    public void concat() throws IOException, ExecutionException, InterruptedException {

        ProcessBuilder ffmpegProcessBuilder = new ProcessBuilder("external/ffmpeg.exe",
                "-protocol_whitelist", "file,pipe,concat",
                "-f", "concat",
                "-safe", "0",
                "-i", "-",
                "-f", "ipod",
                "-c", "copy",
                outputFileName);

        Process ffmpegProcess = ffmpegProcessBuilder.start();
        PrintWriter ffmpegOut = new PrintWriter(ffmpegProcess.getOutputStream());

        StreamCopier ffmpegToOut = new StreamCopier(ffmpegProcess.getInputStream(), NullOutputStream.NULL_OUTPUT_STREAM);
        Future<Long> ffmpegFuture = Executors.newWorkStealingPool().submit(ffmpegToOut);
        StreamCopier ffmpegToErr = new StreamCopier(ffmpegProcess.getErrorStream(), NullOutputStream.NULL_OUTPUT_STREAM);
        Future<Long> ffmpegErrFuture = Executors.newWorkStealingPool().submit(ffmpegToErr);


        for (Future<ConverterOutput> future : futures) {
            ConverterOutput output = future.get();
            ffmpegOut.println("file '" + output.getOutputFileName() + "'");
            ffmpegOut.flush();
        }
        ffmpegOut.close();
        ffmpegFuture.get();

        for (Future<ConverterOutput> future : futures) {
            ConverterOutput output = future.get();
            FileUtils.deleteQuietly(new File(output.getOutputFileName()));
        }

    }
}

File: src/main/scala/uk/yermak/audiobookconverter/FFMpegConcatenator.scala
package uk.yermak.audiobookconverter

import java.io.{File, IOException}
import java.net.URISyntaxException
import java.util.concurrent.TimeUnit

import net.bramp.ffmpeg.progress.{Progress, TcpProgressParser}

/**
  * Created by Yermak on 29-Dec-17.
  */
object FFMpegConcatenator {
  private val FFMPEG = new File("external/x64/ffmpeg.exe").getAbsolutePath
}

class FFMpegConcatenator(var conversion: Conversion, val outputFileName: String, var metaDataFileName: String, var fileListFileName: String, var callback: ProgressCallback) {
  private var progressParser: TcpProgressParser = _

  @throws[IOException]
  @throws[InterruptedException]
  def concat(): Unit = {
    if (conversion.getStatus.isOver) return
    while ( {
      ProgressStatus.PAUSED == conversion.getStatus
    }) Thread.sleep(1000)
    callback.reset()
    try {
      progressParser = new TcpProgressParser((progress: Progress) => callback.converted(progress.out_time_ns / 1000000, progress.total_size))
      progressParser.start()
    } catch {
      case e: URISyntaxException =>
    }
    var process: Process = null
    try {
      val ffmpegProcessBuilder = new ProcessBuilder(FFMpegConcatenator.FFMPEG, "-protocol_whitelist", "file,pipe,concat", "-vn", "-f", "concat", "-safe", "0", "-i", fileListFileName, "-i", metaDataFileName, "-map_metadata", "1", "-f", "ipod", "-c:a", "copy", "-movflags", "+faststart", "-progress", progressParser.getUri.toString, outputFileName)
      process = ffmpegProcessBuilder.start
      StreamCopier.copy(process.getInputStream, System.out)
      StreamCopier.copy(process.getErrorStream, System.err)
      var finished = false
      while ( {
        !conversion.getStatus.isOver && !finished
      }) finished = process.waitFor(500, TimeUnit.MILLISECONDS)
    } finally {
      Utils.closeSilently(process)
      Utils.closeSilently(progressParser)
    }
  }
}